### PR TITLE
ptouch-driver: init at 1.7.0

### PR DESCRIPTION
**Pull request by sascha8a**
## Description of changes

This package provides a printer driver for the Brother PT- and QT- printers. It is based upon the open source repository [https://github.com/philpem/printer-driver-ptouch](https://github.com/philpem/printer-driver-ptouch).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+}:
+
+stdenv.mkDerivation rec {
+  pname = "ptouch-driver";
+  version = "1.7";
+
+  src = fetchFromGitHub {
+    owner = "philpem";
+    repo = "printer-driver-ptouch";
+    rev = "v1.7";
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:25]:
```suggestion
    rev = "v${version}";
```

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    "rastertoptch"  # ptouch-driver
+    "gs"  # ghostscript
+  ];
+
+  meta = with lib; {
+    changelog = "https://github.com/OpenPrinting/foomatic-db/blob/${src.rev}/ChangeLog";
+    description = "Printer Driver for Brother P-touch and QL Label Printers";
+    downloadPage = "https://github.com/philpem/printer-driver-ptouch";
+    homepage = "https://github.com/philpem/printer-driver-ptouch";
+    license = lib.licenses.gpl2Only;
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:83]:
```suggestion
    license = licenses.gpl2Only;
```

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    maintainers = with maintainers; [ sascha8a christoph-heiss ];
+    platforms = lib.platforms.linux;
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:85]:
```suggestion
    platforms = platforms.linux;
```

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  # Comments indicate the respective
+  # package the command is contained in.
+  ppdFileCommands = [
+    "rastertoptch"  # ptouch-driver
+    "gs"  # ghostscript
+  ];
+
+  meta = with lib; {
+    changelog = "https://github.com/OpenPrinting/foomatic-db/blob/${src.rev}/ChangeLog";
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:79]:
```suggestion
    changelog = "https://github.com/philpem/printer-driver-ptouch/releases/tag/v${version}";
```

Probably a copy&paste-leftover?

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    find -H "${placeholder "out"}/share/cups/model/ptouch-driver" -type f -iname '*.ppd' -print0  \
+      | xargs -0r -n 64 -P "$NIX_BUILD_CORES" gzip -9n
+  '';
+
+  # Comments indicate the respective
+  # package the command is contained in.
+  ppdFileCommands = [
+    "rastertoptch"  # ptouch-driver
+    "gs"  # ghostscript
+  ];
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:76]: Based on a quick scan through the files, probably needs `foomatic-rip` from `cups-filters` too?

Don't really have experience with PPD files/PostScript tho.

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  patches = [
+    # Fixes some invalid XML file that otherwise causes a build failure
+    (fetchpatch {
+      name = "fix-brother-ql-600.xml";
+      url = "https://github.com/philpem/printer-driver-ptouch/commit/b3c53b3bc4dd98ed172f2c79405c7c09b3b3836a.patch";
+      hash = "sha256-y5bHKFeRXx8Wdl1++l4QNGgiY41LY5uzrRdOlaZyF9I=";
+    })
+  ];
+
+  buildInputs = [ cups cups-filters foomatic-db foomatic-db-engine ghostscript libpng libxml2 ];
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:38]:
```suggestion
  buildInputs = [ cups ghostscript libpng libxml2 patchPpdFilesHook ];
```

`cups-filters` (see below tho), `foomatic-db` and `foomatic-db-engine` are not needed at runtime, AFAICS? Or is there some non-obvious runtime dependency there?

Also, `patchPpdFilesHook` needs be included here for the hook to be actually executed.

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  nativeBuildInputs = [
+    autoconf
+    automake
+    foomatic-db-engine
+    perl
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:43]:
```suggestion
```

Already included transitively through the below.

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    owner = "philpem";
+    repo = "printer-driver-ptouch";
+    rev = "v1.7";
+    hash = "sha256-3ZotSHn7lERp53hAzx47Ct/k565rEoensCcltwX/Xls=";
+  };
+
+  patches = [
+    # Fixes some invalid XML file that otherwise causes a build failure
+    (fetchpatch {
+      name = "fix-brother-ql-600.xml";
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:32]:
```suggestion
      name = "fix-brother-ql-600.patch";
```

**Review by christoph-heiss — CHANGES_REQUESTED** [review]
A few small nits, plus `buildInputs` can be a bit reduced.

Result of `nixpkgs-review pr 289111` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ptouch-driver</li>
  </ul>
</details>

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 9]
```diff
@@ -0,0 +1,91 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, cups
+, cups-filters
+, foomatic-db
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:6]:
```suggestion
```
Can be dropped, as it is (now) unused.

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    find -H "${placeholder "out"}/share/cups/model/ptouch-driver" -type f -iname '*.ppd' -print0  \
+      | xargs -0r -n 64 -P "$NIX_BUILD_CORES" gzip -9n
+  '';
+
+  # Comments indicate the respective
+  # package the command is contained in.
+  ppdFileCommands = [
+    "rastertoptch" # ptouch-driver
+    "gs" # ghostscript
+    "foomatic-rip" # foomatic-db
```
Comment (christoph-heiss) [inline pkgs/by-name/pt/ptouch-driver/package.nix:75]:
```suggestion
    "foomatic-rip" # cups-filters
```
Since that is where we actually pull it from.

**Review by christoph-heiss — CHANGES_REQUESTED** [review]
Two more small nits, other than that LGTM.

**Comment by Yarny0** [comment]
#### Review per https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#new-packages

##### Reviewed points

- [ ] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on x86-64
- [ ] executables tested on ARCHITECTURE
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [ ] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored


##### Possible improvements

*   Not sure about the license.  The top comment in `rastertoptch.c` claims

    > ptouch-driver is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

    Maybe GPL2+ is more fitting.

*   `patchPpdFilesHook` is only used at build time and should be placed in `nativeBuildInputs`.

*   This package merely creates about 40 ppd files.  I suggest to reduce the `-n` value for `xargs` from `64` to `4` for more parallelism in `postFixup`.

Looks good otherwise.  Thanks for this package!

**Review by christoph-heiss — APPROVED** [review]
Diff & resulting files LGTM.

Result of `nixpkgs-review pr 289111` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ptouch-driver</li>
  </ul>
</details>

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+  nativeBuildInputs = [
+    autoconf
+    automake
+    foomatic-db-engine
+    patchPpdFilesHook
+    (perl.withPackages (pp: with pp; [ XMLLibXML ]))
+  ];
+
+  postPatch = ''
+    patchShebangs .
```
Comment (SuperSandro2000) [inline pkgs/by-name/pt/ptouch-driver/package.nix:47]: Can we narrow this down to some directory?

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+      name = "fix-brother-ql-600.patch";
+      url = "https://github.com/philpem/printer-driver-ptouch/commit/b3c53b3bc4dd98ed172f2c79405c7c09b3b3836a.patch";
+      hash = "sha256-y5bHKFeRXx8Wdl1++l4QNGgiY41LY5uzrRdOlaZyF9I=";
+    })
+  ];
+
+  buildInputs = [ cups cups-filters ghostscript libpng libxml2 ];
+  nativeBuildInputs = [
+    autoconf
+    automake
```
Comment (SuperSandro2000) [inline pkgs/by-name/pt/ptouch-driver/package.nix:40]: Can we use autoreconfHook instead?

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    foomatic-db-engine
+    patchPpdFilesHook
+    (perl.withPackages (pp: with pp; [ XMLLibXML ]))
+  ];
+
+  postPatch = ''
+    patchShebangs .
+  '';
+
+  preConfigure = ''
```
Comment (SuperSandro2000) [inline pkgs/by-name/pt/ptouch-driver/package.nix:50]: And with autoreconfHook we can probably drop this

File: pkgs/by-name/pt/ptouch-driver/package.nix [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    ./autogen.sh
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    make install
```
Comment (SuperSandro2000) [inline pkgs/by-name/pt/ptouch-driver/package.nix:56]: Please use makeTarget to change how make is called and then move the rest into postInstall

File: pkgs/by-name/pt/ptouch-driver/package.nix (hunk 11 above)
Comment (Sascha8a) [reply pkgs/by-name/pt/ptouch-driver/package.nix]: `./foomaticalize` should suffice.

File: pkgs/by-name/pt/ptouch-driver/package.nix (hunk 12 above)
Comment (Sascha8a) [reply pkgs/by-name/pt/ptouch-driver/package.nix]: Done.

File: pkgs/by-name/pt/ptouch-driver/package.nix (hunk 13 above)
Comment (Sascha8a) [reply pkgs/by-name/pt/ptouch-driver/package.nix]: Indeed.

**Pull request merged**